src/core/socket.ts

```diff
@@ -7,21 +7,105 @@ export interface ServerSocketConfig {
 
 export class ServerSocket {
 
-    public server: websocket.server;
+    private server: server;
+    private subscribers: Array<(socket: Socket) => void>;
+    private sockets: Socket[];
 
+    /**
+     * Init socket server
+     * Handle new connections
+     *
+     * @param  {ServerSocketConfig} config
+     */
     constructor(config: ServerSocketConfig) {
         /* istanbul ignore next */
-        const server = http.createServer((request, response) => {
+        const httpServer = http.createServer((request, response) => {
             response.writeHead(404);
             response.end();
         });
-        server.listen(config.port);
-        this.server = new websocket.server({
-            httpServer: server,
+        httpServer.listen(config.port);
+        this.server = new server({
+            httpServer,
             autoAcceptConnections: false
         });
-        // wsServer.on('request', req => {
-        //    this.connection = req.accept('echo-protocol', req.origin);
-        // });
+        this.sockets = [];
+        this.subscribers = [];
+        /* istanbul ignore next */
+        this.server.on('request', req => {
+            const connection = req.accept('echo-protocol', req.origin);
+            const socket = new Socket(req, connection, this);
+            const index = this.sockets.push(socket) - 1;
+            this.subscribers.forEach(sub => sub.apply(this, [ socket ]));
+            connection.on('close', conn => {
+                this.sockets.splice(index, 1);
+            });
+        });
+    }
+
+    /**
+     * Subscribe to new socket connections
+     *
+     * @param  {(socket:Socket)=>void} callback
+     */
+    public onRequest(callback: (socket: Socket) => void) {
+        this.subscribers.push(callback);
+    }
+
+    /**
+     * Get active sockets
+     *
+     * @returns Socket
+     */
+    public getSockets(): Socket[] {
+        return this.sockets;
+    }
+}
+
+export class Socket {
+
+    constructor(
+        private request: request,
+        private connection: connection,
+        private server: ServerSocket
+    ) {}
+
+    /**
+     * Listen events
+     *
+     * @param  {string} event
+     * @param  {()=>void} callback
+     */
+    on(event: string, callback: () => void) {
+        this.connection.on(event, callback);
+    }
+
+    /**
+     * Send data
+     *
+     * @param  {string} event
+     * @param  {any} data
+     */
+    emit(event: string, data: any) {
+        this.connection.sendUTF(JSON.stringify({
+            type: event,
+            data
+        }));
+    }
+
+    /**
```

Comment:
Could you remove the emitAll from the "Socket" definition, cause it's not related to a socket instance but more related to the object which will contains the sockets array